Create and use new `usePagination` hook for paginating loading

mqp opened the pull request:
WIP. This is intended to replace the existing user comments list, user bets list, contract bets list, and maybe other things where we need paginated loading to deal with the quantity of data we are interested in.

Feedback on the general design is welcome. (Feedback on tiny details might be useless as I continue to refactor the thing.)

TODO: General improvements, good pagination UI, fixing the client filter to be smarter (i.e. it should continue to load more pages until you filled up one page size worth of stuff.) Think about react-query integration for making prev-next instantaneous. Think about the matter of live updating on individual items.

jahooma reviewed (commented):
Seems like a pretty good approach! It looks like nearly all the pagination logic for queries can be factored out.

I'm curious how this will work for bets, because we also have to fetch their contracts. But the ordering of the contracts is based on the bets. The best solution might require cached data structures per contract you have bet on that include the summary stats like lastBetTime, profit, invested, etc. And it would be querying this data structure that is paginated with this hook.

jahooma commented on web/hooks/use-pagination.ts:
```diff
@@ -0,0 +1,158 @@
+        }
+      }
+
+      case 'LOAD': {
```
I guess you said react query could be used as this cache. I'm not sure about their API here (don't they only have hooks?), but it sounds more complex than keeping the fetched results in an array. Maybe there's another benefit?

mqp replied:
I kind of agree with you that I think the fetched results should just pile up in the state.

jahooma commented:
Also, IMO, we can punt on live updating the queried items. Bets and comments don't frequently update.

mqp commented:
> I'm curious how this will work for bets, because we also have to fetch their contracts. But the ordering of the contracts is based on the bets. The best solution might require cached data structures per contract you have bet on that include the summary stats like lastBetTime, profit, invested, etc. And it would be querying this data structure that is paginated with this hook.

It's possible the cleanest solution will be for each bet to fetch its own contract in the bet rendering component, but use react-query to have a shared fetch cache between all the bets.

mqp commented:
> I'm curious how this will work for bets, because we also have to fetch their contracts. But the ordering of the contracts is based on the bets. The best solution might require cached data structures per contract you have bet on that include the summary stats like lastBetTime, profit, invested, etc. And it would be querying this data structure that is paginated with this hook.

I'm not 100% sure yet, but I think your idea must be basically where we need to go.

mqp commented:
OK, I did some more work on this:

- Fetched docs now stay in the state so that going back to previous pages doesn't re-fetch.
- Mostly as a result, it got much simpler.
- Fixed up the UI of the controls to look OK.
- Fixed the comment list to use the new `commentType` field to make query more efficient.

This is clearly now the Right Thing for the comments list so I'm going to merge it to get fast comment list and we can consider to use it for fancier things later.
